src/shared/addons/addonreplacer.cpp

```diff
@@ -0,0 +1,149 @@
+  QDir addonPath = manifestFileInfo.dir();
+
+  for (const QJsonValue& urlValue : urls) {
+    QJsonObject urlObj = urlValue.toObject();
+
+    QString urlRequest = urlObj["request"].toString();
+    QString urlResponse = urlObj["response"].toString();
+    QString urlResponseFile(addonPath.filePath(urlResponse));
+
+    if (urlRequest.isEmpty() || urlResponse.isEmpty()) {
+      logger.warning() << "Invalid URL content";
+      return nullptr;
+    }
+
+    QString type = urlObj["type"].toString();
+    if (type.isEmpty() || type == "file") {
+      if (!QFile::exists(urlResponseFile)) {
+        logger.warning() << "The addon does not contain the file `"
+                         << urlResponseFile << "' required for `" << urlRequest
+                         << "'";
+        return nullptr;
+      }
+
+      replacer->m_replaces.append({eUrlFile, urlRequest, urlResponseFile});
+    } else if (type == "directory") {
+      if (!QDir(urlResponseFile).exists()) {
+        logger.warning() << "The addon does not contain the directory `"
+                         << urlResponseFile << "' required for `" << urlRequest
+                         << "'";
+        return nullptr;
+      }
+
+      // To simplify the path comparison, let's add a final '/' in both
+      // request/response strings.
+
+      if (!urlRequest.endsWith('/')) {
+        urlRequest.append('/');
+      }
+
+      if (!urlResponseFile.endsWith('/')) {
+        urlResponseFile.append('/');
+      }
+
+      replacer->m_replaces.append({eUrlDirectory, urlRequest, urlResponseFile});
+    } else {
+      logger.warning() << "Invalid URL type:" << type;
+      return nullptr;
+    }
+  }
+
+  guard.dismiss();
+  return replacer;
+}
+
+AddonReplacer::AddonReplacer(QObject* parent, const QString& manifestFileName,
+                             const QString& id, const QString& name)
+    : Addon(parent, manifestFileName, id, name, "replacer") {
+  MZ_COUNT_CTOR(AddonReplacer);
+}
+
+AddonReplacer::~AddonReplacer() { MZ_COUNT_DTOR(AddonReplacer); }
+
+void AddonReplacer::enable() {
+  Addon::enable();
+  QmlEngineHolder::instance()->engine()->addUrlInterceptor(this);
+  ResourceLoader::instance()->addUrlInterceptor(this);
+}
+
+void AddonReplacer::disable() {
+  Addon::disable();
+  QmlEngineHolder::instance()->engine()->removeUrlInterceptor(this);
+  ResourceLoader::instance()->removeUrlInterceptor(this);
+}
+
+QUrl AddonReplacer::intercept(const QUrl& url,
+                              QQmlAbstractUrlInterceptor::DataType type) {
+  if (type != DataType::QmlFile) {
+    return url;
+  }
+
+  QString urlRequest = url.toString();
+
+  for (const Replace& replace : m_replaces) {
+    if (replace.m_type == eUrlFile) {
+      if (replace.m_request != urlRequest) continue;
+      return normalizeUrlScheme(url, replace.m_response);
+    }
+
+    Q_ASSERT(replace.m_type == eUrlDirectory);
+    if (!urlRequest.startsWith(replace.m_request)) continue;
```

Comment:
Hm, what if we are replacing the folder `qrc:/components/mycomponent` and we are faced with the urlRequest `qrc:/components/mycomponent.qml`. Would that replace the qml file with a folder? Do we need to do a file extension check?

Reply:
no because `mycomponent.qml` is not included in `mycomponent`. There is a test for it.